janus_core/calculations/base.py

```diff
@@ -146,12 +150,23 @@ def __init__(
         self.read_kwargs = read_kwargs
         self.calc_kwargs = calc_kwargs
         self.log_kwargs = log_kwargs
-        self.track_carbon = track_carbon
         self.tracker_kwargs = tracker_kwargs
 
         if not self.model_path and "model_path" in self.calc_kwargs:
             raise ValueError("`model_path` must be passed explicitly")
 
+        if "filename" in log_kwargs:
+            attach_logger = True
+        else:
+            attach_logger = attach_logger if attach_logger else False
+
+        if not attach_logger:
+            if track_carbon:
+                raise ValueError("Carbon tracking requires logging to be enabled")
+            self.track_carbon = False
+        else:
+            self.track_carbon = track_carbon if track_carbon is not None else True
+
```

Comment:
Generally looks good to me.

Is there a utility we could write to handle setting/checking ```attach_logger``` and ```self.track_carbon``` (or ```track_carbon```) in ```training/preprocess```, ```training/train```, and ```BaseCalculation```? Or is that more trouble than its worth.

Reply:
Good point, the training ones are slightly different since they don't operate on `self.logger`, but I can probably pass them through to make them consistent, which would probably be slightly neater.